src/Storages/Streaming/StorageStream.cpp

```diff
@@ -1937,4 +1937,13 @@ void StorageStream::checkReady() const
         throw Exception(
             ErrorCodes::RESOURCE_NOT_INITED, "Background resources of '{}' are initializing", getStorageID().getFullTableName());
 }
+std::vector<nlog::RecordSN> StorageStream::getEveryShardLastSN()
+{
+    std::vector<nlog::RecordSN> lastsn;
+    for (const auto & stream_shard : stream_shards)
+        if (stream_shard->storage)
```

Comment:
why check it ?

Reply:
this interface should always return last sns of all shards.
for local shards, if `last_sn` < 0, means no data
for virtual shards (AKA. remote shards), the `last_sn` always is invalid (< 0)